src/Dfc.CourseDirectory.Web/Controllers/CopyCourse/CopyCourseRunController.cs

```diff
@@ -262,243 +219,355 @@ private async Task<SelectVenueModel> GetVenuesByUkprn(int ukprn)
+                    courseRun.StudyMode = model.StudyMode;
 
-                    courseRuns.Add(copiedCourseRun);
-                    courseForEdit.Value.CourseRuns = courseRuns;
+                    courseRun.Regions = null;
+                    courseRun.VenueId = model.VenueId;
+                    break;
+                case DeliveryMode.WorkBased:
+                    courseRun.VenueId = null;
 
-                    _session.Remove("NewAddedVenue");
-                    _session.Remove("Option");
+                    var availableRegions = new SelectRegionModel();
 
-                    var updatedCourse = await _courseService.UpdateCourseAsync(courseForEdit.Value);
-                    if (updatedCourse.IsSuccess && updatedCourse.HasValue)
+                    if (model.National)
                     {
-                        return RedirectToAction("index", "ProviderCourses",
-                            new
-                            {
-                                notificationTitle = "New course added",
-                                notificationMessage = "You added",
-                                courseRunId = copiedCourseRun.id
-                            });
+                        courseRun.National = true;
+                        courseRun.Regions = availableRegions.RegionItems.Select(x => x.Id).ToList();
                     }
+                    else
+                    {
+                        courseRun.National = false;
+                        courseRun.Regions = model.SelectedRegions;
+                        string[] selectedRegions = availableRegions.SubRegionsDataCleanse(courseRun.Regions.ToList());
 
+                        var subRegions = selectedRegions.Select(selectedRegion => availableRegions.GetSubRegionItemByRegionCode(selectedRegion)).ToList();
+                        courseRun.SubRegions = subRegions;
 
-                }
+                        courseRun.AttendancePattern = AttendancePattern.Undefined;
+                        courseRun.StudyMode = StudyMode.Undefined;
+                    }
+                    break;
+                case DeliveryMode.Online:
+
+                    courseRun.Regions = null;
+                    courseRun.VenueId = null;
+                    courseRun.AttendancePattern = AttendancePattern.Undefined;
+                    courseRun.StudyMode = StudyMode.Undefined;
+                    break;
             }
 
-            return View("Error", new ErrorViewModel { RequestId = Activity.Current?.Id ?? HttpContext.TraceIdentifier });
+            course.Value.CourseRuns = course.Value.CourseRuns.Append(courseRun);
+
+            try
+            {
+                var result = await _courseService.UpdateCourseAsync(course.Value);
+
+                if (!result.IsSuccess || !result.HasValue)
+                {
+                    return View("Error", new ErrorViewModel { RequestId = Activity.Current?.Id ?? HttpContext.TraceIdentifier });
```

Comment:
I think this should throw otherwise we won't see the error in logs

Reply:
I've not actually changed the behaviour of this from the original (just inverted ifs to reduce nesting)